### PR TITLE
restore 1.x behaviour: user docs in conflict cannot login

### PR DESCRIPTION
**Pull request by janl**

### CODE REVIEW

**Comment by janl** [comment]
tests for this will be landing with https://github.com/apache/couchdb/pull/409

File: src/chttpd_auth_cache.erl [hunk 1]
```diff
@@ -209,3 +206,18 @@ update_doc_ignoring_conflict(DbName, Doc, Options) ->
         throw:conflict ->
             ok
     end.
+
+validate_user_creds(nil) ->
+    nil;
+validate_user_creds(UserCreds) ->
+    couch_log:info("~n~n validate_user_creds: ~p~n~n", [UserCreds]),
```
Comment (KlausTrainer) [inline src/chttpd_auth_cache.erl]: No strong opinion, but I'd tend to rather use `debug` log level here.

**Comment by KlausTrainer** [comment]
:+1:

File: src/chttpd_auth_cache.erl (hunk 1 above)
Comment (kxepal) [inline src/chttpd_auth_cache.erl]: Agree. Logging on info will cause information leak though logs and systems they analyse in something more than developer environment.

Comment (kxepal) [inline src/chttpd_auth_cache.erl]: Also, why so much `~n`?

**Comment by kxepal** [comment]
Except logging moment, +1. Nice catch!

File: src/chttpd_auth_cache.erl (hunk 1 above)
Comment (janl) [inline src/chttpd_auth_cache.erl]: oh sorry, this is leftover debugging code, it’ll go away completely

**Comment by rnewson** [comment]
might be good to have a config toggle on this, enabled by default.

**Comment by kxepal** [comment]
@rnewson What's the case when we want to have such behaviour disabled? I don't think user will be happy if they would have specify one password or another depending on which conflict doc wins.

**Comment by rnewson** [comment]
Well, I think we might turn it off for Cloudant so that customers aren't locked out of their accounts but that's a conversation we'll be having this week. Let's not pretend this is a desirable behaviour for users, just the best we can do easily.

**Comment by janl** [comment]
New version without debugging statement and with config option

**Comment by kxepal** [comment]
+1

File: src/chttpd_auth_cache.erl [hunk 2]
```diff
@@ -209,3 +206,28 @@ update_doc_ignoring_conflict(DbName, Doc, Options) ->
         throw:conflict ->
             ok
     end.
+
+maybe_validate_user_creds(nil) ->
+    nil;
+maybe_validate_user_creds(UserCreds) ->
+    AllowConflictedUserDocs = config:get("chttpd_auth", "allow_conflicted_user_docs", "false"),
```
Comment (rnewson) [inline src/chttpd_auth_cache.erl]: use config:get_boolean

File: src/chttpd_auth_cache.erl [hunk 3]
```diff
@@ -209,3 +206,28 @@ update_doc_ignoring_conflict(DbName, Doc, Options) ->
...
+    maybe_validate_user_creds(UserCreds, AllowConflictedUserDocs).
```
Comment (rnewson) [inline src/chttpd_auth_cache.erl]: then `AllowConfictedUserDocs andalso validate_user_creds(UserCreds)`

File: src/chttpd_auth_cache.erl [hunk 4]
```diff
@@ -209,3 +206,28 @@ update_doc_ignoring_conflict(DbName, Doc, Options) ->
         throw:conflict ->
             ok
     end.
+
+maybe_validate_user_creds(nil) ->
```
Comment (rnewson) [inline src/chttpd_auth_cache.erl]: a `-spec` here would help readability, so we know what the possible results are supposed to be.

File: src/chttpd_auth_cache.erl (hunk 3 above)
Comment (rnewson) [inline src/chttpd_auth_cache.erl]: enh, no, because validate_user_creds has an odd contract.

File: src/chttpd_auth_cache.erl [hunk 5]
```diff
@@ -209,3 +206,28 @@ update_doc_ignoring_conflict(DbName, Doc, Options) ->
...
+maybe_validate_user_creds(UserCreds, "false") ->
+    UserCreds;
+maybe_validate_user_creds(UserCreds, "true") ->
+    validate_user_creds(UserCreds).
+
+validate_user_creds(nil) ->
+    nil;
+validate_user_creds(UserCreds) ->
+    case couch_util:get_value(<<"_conflicts">>, UserCreds) of
+    undefined ->
```
Comment (rnewson) [inline src/chttpd_auth_cache.erl]: indentation

**Comment by rnewson** [comment]
I'd at least like to see the get_boolean comment addressed. The odd contract for the result might confound further simplification, though.

**Comment by kxepal** [comment]
Oh, I should take a look on it closer. Thanks, @rnewson! Agree with your notes.

**Comment by janl** [comment]
new version, uses `get_boolean`, added comment to clarify function behaviour

skipped spec because of odd function sig.

**Comment by janl** [comment]
latest version should fix travis issues.

**Comment by janl** [comment]
voilá

**Comment by rnewson** [comment]
+1

**Comment by janl** [comment]
thanks all!

**Pull request merged**